File: server/src/combat/attr_set.cpp
#include "attr_set.h"
#include "math_utils.h"
#include "bit_stream.h"

#include <algorithm>
#include <spdlog/spdlog.h>

void AttrSet::net_serialize(OutputBitStream& bs) {
    bs.write(_max_health);
    bs.write(_health);
    bs.write(_max_mana);
    bs.write(_mana);
}

bool AttrSet::consume_dirty(OutputBitStream& bs) {
    bool dirty = false;
    bs.write(_dirty_flag);
    if (_dirty_flag) {
        dirty = true;

        if (_dirty_flag & (uint16_t)DirtyFlag::max_health) {
            bs.write(_max_health);
        }

        if (_dirty_flag & (uint16_t)DirtyFlag::health) {
            bs.write(_health);
        }

        if (_dirty_flag & (uint16_t)DirtyFlag::max_mana) {
            bs.write(_max_mana);
        }

        if (_dirty_flag & (uint16_t)DirtyFlag::mana) {
            bs.write(_mana);
        }

        _dirty_flag = 0;
    }
    return dirty;
}

void AttrSet::add_health(int val) {
    int n = _health + val;
    n = std::max(0, std::min(_max_health, n));
    set_health(n);
}

void AttrSet::add_mana(int val) {
    int n = _mana + val;
    n = std::max(0, std::min(_max_mana, n));
    set_mana(n);
}

int AttrSet::take_damage(AttrSet& attacker, int attack_damage) {
    // 1. 基础伤害计算（至少为1）
    int base_damage = (std::max)(1, attack_damage - this->defence);

    // 2. 暴击判断
    bool is_critical = random_01() < attacker.critical_rate;
    int damage = is_critical ?
        static_cast<int>(base_damage * (1.0f + attacker.critical_damage)) : base_damage;

    // 3. 命中与闪避判断
    float hit_chance = attacker.accuracy - this->dodge_rate;
    bool is_hit = random_01() < hit_chance;
    if (!is_hit) {
        damage = 0;
    }

    // 4. 护盾吸收
    int shield_absorb = (std::min)(damage, this->shield);
    this->shield -= shield_absorb;
    damage -= shield_absorb;

    // 5. 实际扣血（确保不小于0）
    add_health(-damage);

    // 6. 吸血效果（不超过最大生命值）
    if (damage > 0 && attacker.lifesteal > 0) {
        int heal_amount = static_cast<int>(damage * attacker.lifesteal);
        attacker.add_health(heal_amount);
    }

    return damage;
}

void AttrSet::clamp_all_attributes() {
    accuracy = std::clamp(accuracy, 0.0f, 1.0f);
    dodge_rate = std::clamp(dodge_rate, 0.0f, 1.0f);
    tenacity = std::clamp(tenacity, 0.0f, 1.0f);
}

File: client/Assets/Scripts/Combat/AttrSet.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class AttrSet
{
    public int MaxHealth;
    public int Health;
    public int MaxMana;
    public int Mana;
    public int Attack;
    public float AttackSpeed = 1;
    public float AdditionalAttackSpeed = 0;

    public int Defence;
    public float CriticalRate;
    public float CriticalDamage;
    // 命中率
    [Range(0.0f, 1.0f)]
    public float Accuracy;
    // 闪避
    [Range(0.0f, 1.0f)]
    public float DodgeRate;
    // 护盾
    public int Shield;
    // 韧性
    [Range(0.0f, 1.0f)]
    public float Tenacity;
    // 吸血
    [Range(0.0f, 1.0f)]
    public float Lifesteal;

    // 状态（晕眩、冻结、霸体。。。）
    public int Status;

    public void TakeDamage(AttrSet attacker, int attackDamage)
    {
        // 1. 基础伤害计算
        int baseDamage = Mathf.Max(1, attackDamage - this.Defence);

        // 2. 暴击判断
        bool isCritical = Random.value < attacker.CriticalRate;
        float damage = isCritical ? baseDamage * (1 + attacker.CriticalDamage) : baseDamage;

        // 3. 命中与闪避判断
        bool isHit = Random.value < (attacker.Accuracy - this.DodgeRate);
        if (!isHit)
        {
            Debug.Log("攻击未命中！");
            damage = 0;
        }

        // 4. 护盾吸收
        int shieldAbsorb = Mathf.Min((int)damage, this.Shield);
        this.Shield -= shieldAbsorb;
        damage -= shieldAbsorb;

        // 5. 实际扣血
        this.Health -= (int)damage;
        this.Health = Mathf.Max(0, this.Health); // 确保生命值不低于0

        // 6. 吸血效果
        if (damage > 0 && attacker.Lifesteal > 0)
        {
            int healAmount = (int)(damage * attacker.Lifesteal);
            attacker.Health += healAmount;
            attacker.Health = Mathf.Min(attacker.MaxHealth, attacker.Health); // 确保生命值不超过上限
        }

        // 输出日志
        Debug.Log($"攻击者造成伤害: {damage}, 是否暴击: {isCritical}, 护盾吸收: {shieldAbsorb}, 防御者剩余生命: {this.Health}");
    }

}
